aexpect/remote.py

```diff
@@ -469,7 +471,7 @@ def remote_scp(command, password_list, log_filename=None, transfer_timeout=600,
     logging.debug("Trying to SCP with command '%s', timeout %ss",
                   command, transfer_timeout)
     if log_filename:
-        output_func = utils_misc.log_line
+        output_func = log_function
```

Comment:
Well this changes the behaviour as one might specify the `log_filename` as well as `log_function` and expect it to be logged on both places. Similarly setting the `log_filename` and without `log_function` would raise an exception, while before it did exactly what is in docstring.

If you really need to specify the log_function I'd suggest turning the `log_filename` into 3-state:

* str - write to file/log/stream
* callable - call it
* None - do nothing

alternatively we could keep both of them, but then the docstring has to mention they are mutually exclusive and when `log_filename` is specified it should use the internal implementation.

Reply:
PS: I haven't tested it so I could be wrong...

Reply:
One more thing I noticed, you need to add the arguments as last, otherwise it'd break existing scripts that use positional arguments. Another reason for 3-state.

Let me summarize. I'd prefer:

1. not to introduce log_function
2. introduce it as tri-state via `remote_scp(command, password_list, log_filename=None, transfer_timeout=600, login_timeout=300, log=None):` where `log_filename` would be deprecated, log warning on use and `log` would accept the 3 types
3. introduce the `log_function` but keep backward compatibility: `remote_scp(command, password_list, log_filename=None, transfer_timeout=600, login_timeout=300, log_function=None):`, using `log_filename` and `log_function` should raise exception and `log_filename` should use internal implementatio nof `log_line`.

Reply:
I see, not providing the `log_function` explicitly would imply that we have to provide it myself as before and migrate some code from `utils_misc` as well. I am against tristates or other magical multitypes which can lead to a lot of future headache. The third option is also plausible but first I will look into a way to port the additional code like `utils_misc.get_log_filename` and `utils_misc.log_line`.

Reply:
Hmmm, one major reason I left the log function implementation to the caller is that `utils_misc` is doing some strange magic with its implementation of `log_line` making use of globals and lots of things that will both complicate the configuration of aexpect and require heavier migration on the side of all avocado-vt. Providing just `log_file` xor `log_line` is also not possible since the aexpect side makes use of both in the implementation of `remote_scp` (thus raising exception if both are supplied no longer makes sense either). This leaves us only with part of your option 3 as a possibility. Any ideas for minimal and easily replaceable (e.g. by avocado-vt) internal `log_line` implementation so that we could default to option 1?